### ISSUE TITLE
ImmutableSortedMap.copyOf and ImmutableSortedSet.copyOf are inconsistent

### ISSUE THREAD

**Issue by simonmcooper** [google/guava#2256]
`ImmutableSortedMap.copyOf` has an overload with the following signature:

    copyOf(Map<? extends K,? extends V> map, Comparator<? super K> comparator);

but similar methods on `ImmutableSortedSet` are declared like so:

    copyOf(Comparator<? super E> comparator, Collection<? extends E> elements)
    copyOf(Comparator<? super E> comparator, Iterable<? extends E> elements)
    copyOf(Comparator<? super E> comparator, Iterator<? extends E> elements)

The placement of the `Comparator` argument is inconsistent. I suggest one version is deprecated, and overloads with the arguments swapped over are added.

**Comment by cpovirk** [comment]
:(

This is sad, though probably not sad enough for the churn that the fix would produce.

If we do want to do something here, I think we want the `Comparator` to come second, since it's more or less an "optional parameter."

**Comment by simonmcooper** [comment]
The `Comparator` coming second also matches with the JCL (`Collections.sort`, etc)

I would say there's no harm in deprecating the existing inconsistent ones, providing overloads, and leaving it like that for several versions. Gives people plenty of time to change.

**Comment by kluever** [comment]
There's only O(hundreds) of usages in google, so this change isn't necessarily impossible. It would be a relatively safe deprecation + migration, unless someone decided to implement both Comparator and Iterable on *both* of the objects (*gasp*).

`ImmutableSortedSet` isn't `@Beta`, so we'd need to do the 2 year deprecation cycle. It's still not out of the question though.

**Comment by kluever** [comment]
We discussed this at API review, and while we acknowledge that this was absolutely a mistake (in the purest design sense), it's probably not worth the churn and potential problems to our open source users to fix it at this time.

If we did change it, most users would just get annoyed at us for making them swap the param ordering (they wouldn't necessarily realize it was to correct a mistake). Also, even given our generous 2 year deprecation cycle for non-@Beta APIs, we still think it might cause version compatibility issues for our open source users.

Given that the incorrect param ordering isn't hurting anything (it's more or less impossible to get the order "wrong"), we're just going to live with this inconsistency. Feel free to cite this issue as a known Guava API mistake if ever asked :-)
